tests/test_hoist.py

```diff
@@ -0,0 +1,264 @@
+
+write_transformation_output = False
+
+
+@pytest.fixture(scope='module', name='here')
+def fixture_here():
+    return Path(__file__).parent
+
+
+@pytest.fixture(name='config')
+def fixture_config():
+    """
+    Default configuration dict with basic options.
+    """
+    return {
+        'default': {
+            'mode': 'idem',
+            'role': 'kernel',
+            'expand': True,
+            'strict': True,
+        },
+        'routines': []
+    }
+
+
+def compile_and_test(scheduler, here, a=(5,)):
+    """
+    Compile the source code and call the driver function in order to test the results for correctness.
+    """
+    assert is_iterable(a) and all(isinstance(_a, int) for _a in a)
+    for item in scheduler.items:
+        if "driver" in item.name:
+            suffix = '.F90'
+            module = jit_compile(item.source, filepath=item.source.path.with_suffix(suffix).name, objname=None)
+            for _a in a:
+                parameter_length = 3
+                b = np.zeros((_a,), dtype=np.int32, order='F')
+                c = np.zeros((_a, parameter_length), dtype=np.int32, order='F')
+                module.Transformation_Module_Hoist.driver(_a, b, c)
+                assert (b == 42).all()
+                assert (c == 11).all()
+            clean_test(filepath=here.parent / item.source.path.with_suffix(suffix).name)
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_hoist(here, frontend, config):
+    """
+    Basic testing of the non-modified Hoist functionality, thus hoisting all (non-parameter) local variables.
+    """
+
+    proj = here/'sources/projHoist'
+
+    config['routine'] = [
+        {
+            'name': 'driver',
+            'role': 'driver',
+            'expand': True,
+        },
+    ]
+
+    scheduler = Scheduler(paths=[proj], config=config, seed_routines=['driver'], frontend=frontend)
+
+    compile_and_test(scheduler=scheduler, here=here, a=(5, 10, 100))
+
+    # Analysis
+    scheduler.process(transformation=HoistAnalysis(), reverse=True)
+    # Synthesis
+    scheduler.process(transformation=HoistSynthesis())
+
+    for item in scheduler.items:
+        if write_transformation_output:
+            suffix = '.out.F90'
+            sourcefile = item.source
+            sourcefile.write(path=sourcefile.path.with_suffix(suffix).name)
+        if "driver" in item.name:
+            assert len(item.routine.arguments) == 3
+        elif "kernel1" in item.name:
+            assert len(item.routine.arguments) == 6
+        elif "kernel2" in item.name:
+            assert len(item.routine.arguments) == 10
+        elif "device1" in item.name:
+            assert len(item.routine.arguments) == 8
+        elif "device2" in item.name:
+            assert len(item.routine.arguments) == 5
+
+    compile_and_test(scheduler=scheduler, here=here, a=(5, 10, 100))
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_hoist_find_arrays(here, frontend, config):
```

Comment:
Genuine question what the right pattern is that we would want for that:
Should this specialization method be done via derived classes that overwrite the `find_variables` method, or should the criteria be formulated as a callback, similar to the [`ExpressionFinder`](https://github.com/ecmwf-ifs/loki/blob/main/loki/expression/expr_visitors.py#L61) visitor?